onecodex/stats.py

```diff
@@ -0,0 +1,235 @@
+            Metadata variable to group samples by. At least two groups are required. If `group_by`
+            is a tuple or list, field values are joined with an underscore character ("_").
+        metric : {'braycurtis', 'jaccard', 'cityblock', 'manhattan', 'aitchison', 'unweighted_unifrac', 'weighted_unifrac'}, optional
+            The beta diversity distance metric to calculate. Note that 'cityblock' and 'manhattan'
+            are equivalent metrics.
+        rank : {'auto', 'kingdom', 'phylum', 'class', 'order', 'family', 'genus', 'species'}, optional
+            Analysis will be restricted to abundances of taxa at the specified level.
+        alpha : float, optional
+            Threshold to determine statistical significance (e.g. p < `alpha`). Must be between 0
+            and 1 (exclusive). If the p-value is significant and there are more than two groups,
+            posthoc pairwise PERMANOVA tests are performed.
+        num_permutations : int, optional
+            Number of permutations to use when computing the p-value.
+
+        Returns
+        -------
+        BetaDiversityStatsResults
+            A dataclass with these attributes:
+            - `test`: stats test that was performed
+            - `statistic`: PERMANOVA pseudo-F test statistic
+            - `pvalue`: p-value based on `num_permutations`
+            - `num_permutations`: number of permutations used to compute `pvalue`
+            - `sample_size`: number of samples used in the test after filtering
+            - `group_by_variable`: name of the variable used to group samples by
+            - `groups`: names of the groups defined by `group_by_variable`
+            - `posthoc`: A dataclass with these attributes (if posthoc results were computed):
+              - `statistics`: `pd.DataFrame` containing pairwise PERMANOVA pseudo-F statistics. The
+                index and columns are sorted group names.
+              - `pvalues`: `pd.DataFrame` containing pairwise PERMANOVA *unadjusted* p-values. The
+                index and columns are sorted group names.
+              - `adjusted_pvalues`: `pd.DataFrame` containing pairwise PERMANOVA *adjusted*
+                p-values. False discovery rate is controlled using Benjamini-Hochberg. The index and
+                columns are sorted group names.
+
+        See Also
+        --------
+        skbio.stats.distance.permanova
+        scipy.stats.false_discovery_control
+
+        """
+        if not (0.0 < alpha < 1.0):
+            raise StatsException("`alpha` must be between 0 and 1 (exclusive).")
+
+        group_by = self._tuplize(group_by)
+
+        # Munge the metadata first in case there's any errors
+        df, magic_fields = self._metadata_fetch([group_by], coerce_missing_composite_fields=False)
+        group_by_column_name = magic_fields[group_by]
+
+        # Compute beta diversity distance matrix
+        dm = self.beta_diversity(metric=metric, rank=rank)
```

Comment:
Nit: we could probably move this test lower so that we don't run it if we know there aren't enough groups?